src/main/java/input_feed/serial/TextToSerialR1CS.java

```diff
@@ -0,0 +1,147 @@
+            BufferedReader brB = new BufferedReader(new FileReader(filePath + ".b"));
+            BufferedReader brC = new BufferedReader(new FileReader(filePath + ".c"));
+
+            for (int currRow = 0; currRow < numConstraints; currRow++){
+                LinearCombination<FieldT> A = makeRowAt(currRow, brA, false);
+                LinearCombination<FieldT> B = makeRowAt(currRow, brB, false);
+                LinearCombination<FieldT> C = makeRowAt(currRow, brC, negateCMatrix);
+
+                constraints.add(new R1CSConstraint<>(A, B, C));
+            }
+            brA.close();
+            brB.close();
+            brC.close();
+
+        } catch (Exception e){
+            System.err.println("Error: " + e.getMessage());
+        }
+
+        return new R1CSRelation<>(constraints, numInputs, numAuxiliary);
+    }
+
+    public Tuple2<Assignment<FieldT>, Assignment<FieldT>> loadWitness() {
+
+        Assignment<FieldT> primary = new Assignment<>();
+        Assignment<FieldT> auxiliary = new Assignment<>();
+
+        try{
+            String[] constraintParameters = new BufferedReader(
+                    new FileReader(filePath + ".problem_size")).readLine().split(" ");
+
+            int numPrimary = Integer.parseInt(constraintParameters[0]);
+            int numAuxiliary = Integer.parseInt(constraintParameters[1]);
+
+            final Assignment<FieldT> fullAssignment = new Assignment<>();
+
+            BufferedReader brA = new BufferedReader(new FileReader(filePath + ".aux"));
+            int count = 0;
+            String nextLine;
+            while ((nextLine = brA.readLine()) != null) {
+                final FieldT value = fieldParameters.construct(nextLine);
+                fullAssignment.add(value);
+                count++;
+            }
+            brA.close();
+
+            BufferedReader brP = new BufferedReader(new FileReader(filePath + ".public"));
+            while ((nextLine = brP.readLine()) != null) {
+                final FieldT value = fieldParameters.construct(nextLine);
+                fullAssignment.add(value);
+                count++;
+            }
+            brP.close();
+
+            assert (count == numPrimary + numAuxiliary);
+
+//            auxiliary = new Assignment<>(fullAssignment.subList(0, numAuxiliary));
+//            primary = new Assignment<>(fullAssignment.subList(numAuxiliary, fullAssignment.size()));
+            primary = new Assignment<>(fullAssignment.subList(0, numPrimary));
+            auxiliary = new Assignment<>(fullAssignment.subList(numPrimary, fullAssignment.size()));
+        } catch (Exception e){
+            System.err.println("Error: " + e.getMessage());
+        }
+        return new Tuple2<>(primary, auxiliary);
+    }
+
+    private LinearCombination<FieldT> makeRowAt (long index, BufferedReader reader, boolean negate) {
+        final LinearCombination<FieldT> L = new LinearCombination<>();
+
+        try {
```

Comment:
It might be worthwhile doing a similar map, reduce style operations as for the distributed reader (1st map all lines to constraintIndex, linear term, then reduce all linear terms with the same constraint index into the linear combination.